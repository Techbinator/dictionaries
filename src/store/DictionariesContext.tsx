import React from "react";
import {
  IDictionaryProvider,
  IDictionaries,
  ITranslation,
  ITranslations
} from "./Dictionaries.type";
import { v1 } from "uuid";

export const DictionariesContext = React.createContext({});
export class DictionariesProvider extends React.Component {
  addDictionary = (name: string) => {
    const uuid = v1();
    const newDictionary = {
      uuid,
      name,
      translationsUUID: []
    };
    this.setState({
      dictionaries: { [uuid]: newDictionary, ...this.state.dictionaries }
    });
  };

  activateDictionary = (dictionaryUUID: string) => {
    this.setState({ activeDictionary: dictionaryUUID });
  };

  removeDictionary = (uuid: string) => {
    const { activeDictionary, dictionaries, translations } = this.state;
    const dictionary = dictionaries[uuid];
    const newDictionaries: IDictionaries = {
      ...dictionaries
    };
    const NewTranslations = {
      ...translations
    };
    dictionary.translationsUUID.forEach(uuid => delete NewTranslations[uuid]);

    delete newDictionaries[uuid];
    this.setState({
      dictionaries: newDictionaries,
      activeDictionary: activeDictionary === uuid ? "" : activeDictionary,
      translations: NewTranslations
    });
  };

  removeTranslation = (dictionaryUUID: string, translationUUID: string) => {
    const { dictionaries, translations } = this.state;
    const newTranslations: ITranslations = {
      ...translations
    };
    delete newTranslations[translationUUID];
    this.setState({
      translations: newTranslations,
      dictionaries: {
        ...dictionaries,
        [dictionaryUUID]: {
          ...dictionaries[dictionaryUUID],
          translationsUUID: dictionaries[
            dictionaryUUID
          ].translationsUUID.filter(tuuid => tuuid !== translationUUID)
        }
      }
    });
  };

  addTranslation = (dictionaryUUID: string, translation: ITranslation) => {
    const { dictionaries, translations } = this.state;
    const uuid = v1();
    const newTranslation = {
      ...translation,
      uuid
    };

    const dictionaryTranslations = dictionaries[dictionaryUUID];
    this.setState({
      translations: {
        [uuid]: newTranslation,
        ...translations
      },
      dictionaries: {
        ...dictionaries,
        [dictionaryUUID]: {
          ...dictionaryTranslations,
          translationsUUID: [uuid, ...dictionaryTranslations.translationsUUID]
        }
      }
    });
  };

  editTranslation = (translation: ITranslation) => {
    this.setState({
      translations: {
        ...this.state.translations,
        [translation.uuid]: translation
      }
    });
  };

  state: IDictionaryProvider = {
    dictionaries: {},
    translations: {},
    activeDictionary: "",
    addDictionary: this.addDictionary,
    removeDictionary: this.removeDictionary,
    removeTranslation: this.removeTranslation,
    addTranslation: this.addTranslation,
    editTranslation: this.editTranslation,
    activateDictionary: this.activateDictionary
  };

  //get from localstorage
  componentWillMount() {
    const dictionaries = localStorage.getItem("dictionaries");
    if (dictionaries) {
      this.setState({ dictionaries: JSON.parse(dictionaries) });
    }

    const translations = localStorage.getItem("translations");
    if (translations) {
      this.setState({ translations: JSON.parse(translations) });
    }
  }
  //save to localstorage
  componentDidUpdate({}, prevState: IDictionaryProvider) {
    const { dictionaries, translations } = this.state;
    if (
      JSON.stringify(dictionaries) !== JSON.stringify(prevState.dictionaries)
    ) {
      // save to localstorage
      localStorage.setItem("dictionaries", JSON.stringify(dictionaries));
    }
    if (
      JSON.stringify(translations) !== JSON.stringify(prevState.translations)
    ) {
      // save to localstorage
      localStorage.setItem("translations", JSON.stringify(translations));
    }
  }

  render() {
    return (
      <DictionariesContext.Provider value={this.state}>
        {this.props.children}
      </DictionariesContext.Provider>
    );
  }
}
