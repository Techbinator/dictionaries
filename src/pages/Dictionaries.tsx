import React from "react";

import Dictionary from "../components/Dictionary";
import NewDictionary from "../components/NewDictionary";
import { DictionariesContext } from "../store/DictionariesContext";

export default class Dictionaries extends React.Component {
  render() {
    const {
      dictionaries,
      translations,
      addDictionary,
      addTranslation,
      removeDictionary,
      removeTranslation,
      editTranslation
    } = this.context;
    return (
      <table className="one-edge-shadow">
        <thead>
          <tr>
            <th colSpan={2}>Name</th>
            <th>Action</th>
          </tr>
        </thead>
        <tbody>
          <NewDictionary addDictionary={addDictionary} />
          {Object.keys(dictionaries).map((keyName: string) => (
            <Dictionary
              key={dictionaries[keyName].uuid}
              dictionary={dictionaries[keyName]}
              translations={translations}
              addTranslation={addTranslation}
              removeDictionary={removeDictionary}
              removeTranslation={removeTranslation}
              editTranslation={editTranslation}
            />
          ))}
        </tbody>
      </table>
    );
  }
}
Dictionaries.contextType = DictionariesContext;
