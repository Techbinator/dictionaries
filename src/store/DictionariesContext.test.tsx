import React from "react";
import { DictionariesProvider } from "./DictionariesContext";
import { shallow } from "enzyme";
import { IDictionaryProvider } from "./Dictionaries.type";

const dummyUUID = "11111-22222-33333-44444";
jest.mock("uuid", () => ({
  v1: jest.fn(() => "11111-22222-33333-44444")
}));

describe("Header component", () => {
  const wrapper: any = shallow(<DictionariesProvider />);
  let state: IDictionaryProvider;
  beforeEach(() => {
    state = wrapper.state();
  });
  it("should render snapshot with default values", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should have correct default state", () => {
    expect(state.dictionaries).toEqual({});
    expect(state.translations).toEqual({});
    expect(state.activeDictionary).toEqual("");
  });

  it("should add dictionary", () => {
    state.addDictionary("test");
    const expected = {
      [dummyUUID]: {
        name: "test",
        translationsUUID: [],
        uuid: dummyUUID
      }
    };
    expect(wrapper.instance().state.dictionaries).toEqual(expected);
  });
  // it("should set item to localstorage", () => {

  //   const localstorageMock = jest.spyOn(
  //     window.localStorage.__proto__,
  //     "setItem"
  //   );
  //   expect(localstorageMock).toHaveBeenCalledWith(
  //     "translations",
  //     '{"11111-22222-33333-44444":{"uuid":"11111-22222-33333-44444","domain":"domainDummyEdited","range":"rangeDummyEdited"}}'
  //   );
  // });
  it("should activate dictionary", () => {
    state.activateDictionary(dummyUUID);
    expect(wrapper.instance().state.activeDictionary).toEqual(dummyUUID);
  });

  it("should add translation", () => {
    state.addTranslation(dummyUUID, {
      uuid: "",
      domain: "domainDummy",
      range: "rangeDummy"
    });
    const expected = {
      [dummyUUID]: {
        uuid: dummyUUID,
        domain: "domainDummy",
        range: "rangeDummy"
      }
    };
    expect(wrapper.instance().state.translations).toEqual(expected);
    expect(
      wrapper.instance().state.dictionaries[dummyUUID].translationsUUID
    ).toContain(dummyUUID);
  });

  it("should edit translation", () => {
    state.editTranslation({
      uuid: dummyUUID,
      domain: "domainDummyEdited",
      range: "rangeDummyEdited"
    });
    const expected = {
      [dummyUUID]: {
        uuid: dummyUUID,
        domain: "domainDummyEdited",
        range: "rangeDummyEdited"
      }
    };
    expect(wrapper.instance().state.translations).toEqual(expected);
  });

  it("should remove translation", () => {
    state.removeTranslation(dummyUUID, dummyUUID);
    const expected = {};
    expect(wrapper.instance().state.translations).toEqual(expected);
    expect(
      wrapper.instance().state.dictionaries[dummyUUID].translationsUUID
    ).toEqual([]);
  });

  it("should remove dictionary", () => {
    state.removeDictionary(dummyUUID);

    expect(wrapper.instance().state.dictionaries).toEqual({});
    expect(wrapper.instance().state.translations).toEqual({});
    expect(wrapper.instance().state.activeDictionary).toEqual("");
  });
});
