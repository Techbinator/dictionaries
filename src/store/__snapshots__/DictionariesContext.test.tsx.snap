// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render snapshot with default values 1`] = `
<ContextProvider
  value={
    Object {
      "activateDictionary": [Function],
      "activeDictionary": "",
      "addDictionary": [Function],
      "addTranslation": [Function],
      "dictionaries": Object {},
      "editTranslation": [Function],
      "removeDictionary": [Function],
      "removeTranslation": [Function],
      "translations": Object {},
    }
  }
/>
`;
